Unified QR code, BIP21 with lightning fallback

rockstardev opened the pull request:
Resolves: #1064 

Here is feature in action:

![bip21 fallback](https://user-images.githubusercontent.com/5191402/98511919-1703df80-222b-11eb-8d8a-5d1a5adfd470.gif)

As part of PR I've also changed lightning invoices to be lowercase by default, this one is long overdue.

shesek commented:
Assuming the QR library you're using supports automatic mixed modes (I don't know if it does), making the bolt11 part of the URI in all-uppercase will take advantage of the more efficient alphanumeric encoding mode, making the QR smaller, less dense and easier to scan from a distance.

There's some information on this on Bitcoin Optech, in the context of bech32 QRs:
https://bitcoinops.org/en/bech32-sending-support/#creating-more-efficient-qr-codes-with-bech32-addresses

> By convention ln invoices are usually lowercase

They usually are when presented as text, but if you're aware of wallets that use lowercase in QRs, it would be good to give them a nudge and tell them to uppercase them :)

rockstardev commented:
Thanks @shesek, I've left lowercase as default now, but tweaked our code to make lninvoice uppercased for QR code. Example QR:

![image](https://user-images.githubusercontent.com/5191402/98515389-9d6ef000-2230-11eb-9715-367b2f80d747.png)

`bitcoin:bcrt1qudfg7xmh47s33gja2mgrpxkt0c2yujyvm6v82e?amount=0.00032589&lightning=LNBCRT325890N1P063ASVPP57NSJTMJUNK09AJAVQ593DHMRFDTHN3QSCGYU3XXA22MQN28U03QQDP62PSKJEPQW3HJQ4MPD3KX2APQVDEX2CT5D9HKUGPGFAEXGETJYPY5GW3Q9YXQZUPCQP2SP5JFUTDSL7Z2S7C2N9TY4YQDY0FFXGHMDF0G47UP402PFYM08ET6NS9QY9QSQRUHPGHUZ66FDKUJ0KK27C50TENA7MDSVV8RMJN4M0KR8SNTDNMHS53NL0HR4DRQ47Z2XSFXT54NRLSF0AH90YWWQQECRGKE6Q687KCQQQE4YJ3`

I was thinking about making `LIGHTNING=` part of querystring uppercase, but I'll wait for wallet devs (@bitbug42 mentioned Eclair supports this lightning fallback) to confirm uppercase would work. Same for `bitcoin:` over `BITCOIN:`

shesek commented:
Have you tried comparing the QRs to see if the uppercase really enables the alphanumeric mode and makes them smaller/sparser?

Upper-casing the query-string parameter name is indeed problematic, but using `BITCOIN:` as the schema should be working, I've been using it myself with no issues. It's also recommended on the Bitcoin Optech article linked above:

> Although the preferred form of BIP21’s scheme identifier `bitcoin:` is lowercase, it can also be uppercased as allowed by both BIP21 and [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1). This also produces a less complex image.

You should also upper-case the bech32 address to take full advantage of the alphanumeric encoding mode (it'll use multiple segments and switch modes between them - an alphanumeric segment for the schema+address, followed by a binary segment for the query, followed by another alphanumeric segment for the bolt11).

Kukks reviewed (dismissed):
Code is good, would have been better to do the cleanup as a separate PR but it's ok. I think I will propose to move the actual feature code to some system plugin in the future.

You need to add the new prop to the swagger docs

rockstardev commented:
@shesek: OK, we've previously removed uppercasing schema in QR Code because some readers were detecting it like `bitcoIN:` for some reason. Here is example QR:

![image](https://user-images.githubusercontent.com/5191402/98633907-fa79ad00-22e7-11eb-9ecd-3deb4cdfe3f8.png)

I've also added uppercasing of receiving address if it starts with bech32 prefixes. Let's see if @NicolasDorier is OK with merging all this.

@kukks: I still can't believe we're manually editing swagger files ;). Added it.

Kukks commented:
> @shesek: OK, we've previously removed uppercasing schema in QR Code because some readers were detecting it like `bitcoIN:` for some reason. Here is example QR:
> 
> ![image](https://user-images.githubusercontent.com/5191402/98633907-fa79ad00-22e7-11eb-9ecd-3deb4cdfe3f8.png)
> 
> I've also added uppercasing of receiving address if it starts with bech32 prefixes. Let's see if @NicolasDorier is OK with merging all this.
> 
> @Kukks: I still can't believe we're manually editing swagger files ;). Added it.

Me too. Feel free to voice your swagger feedback at  #2043

Kukks commented on BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:
```diff
@@ -46,18 +46,41 @@ class Prepare
+                );
+            }
         }
+        private static string[] bech32Prefixes = new[] { "bc1", "tb1", "bcrt1" };
```
We should really not have this hardcoded in here.  You should get this based on the current network using soemthing like `NBitcoin.Network.Main.GetBech32Encoder(Bech32Type.WITNESS_PUBKEY_ADDRESS, false).HumanReadablePart`

NicolasDorier replied:
Can we remove this uppercase stuff?

https://github.com/btcpayserver/btcpayserver/pull/2060/files?w=1#r520458592

I am not comfortable with it. Or maybe we can just activate it for people checking `OnChainWithLnInvoiceFallback` so we can get feedback from early users.

rockstardev replied:
@Kukks these three values are highly unlikely to change, and all the extra code and converting byte[] to string seems unnecessary to me. 

We could do three calls to fetch prefixes from Network class... but this is also why I've added test - so if bech32 prefixes change and world starts to end, we'll get notified that QR codes in BtcPayServer are not properly uppercased ;)

NicolasDorier commented on BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:
```diff
@@ -46,18 +46,41 @@ class Prepare
+            // We're trying to make as many characters uppercase to make QR smaller
+            // Ref: https://github.com/btcpayserver/btcpayserver/pull/2060#issuecomment-723828348
+            model.InvoiceBitcoinUrlQR = cryptoInfo.PaymentUrls.BIP21
+                .Replace("bitcoin:", "BITCOIN:", StringComparison.OrdinalIgnoreCase)
```
I prefer not doing this, as I don't know if various wallet support it.
Lightning is different, because everybody using lightning did it from beginning.

rockstardev replied:
https://github.com/btcpayserver/btcpayserver/pull/2060#issuecomment-727144821

shesek commented:
> @shesek: OK, we've previously removed uppercasing schema in QR Code because some readers were detecting it like bitcoIN: for some reason.

This could show up like this in some QRs readers, but this shouldn't affect the ability to scan it in any way.

> I prefer not doing this, as I don't know if various wallet support it. Lightning is different, because everybody using lightning did it from beginning.

Well, FWIW, I've been using uppercase bech32 addresses in a bunch of my projects with no issues, including in Esplora and Spark Wallet.

And as mentioned before, using an upper case schema is allowed by BIP21, RFC3986, and is recommended by Optech.

bolatovumar commented on BTCPayServer.Client/Models/StoreBaseData.cs:
```diff
@@ -30,31 +30,36 @@ public abstract class StoreBaseData
         public bool AnyoneCanCreateInvoice { get; set; }
 
+
+        //
```
Did you mean to add a comment here but forgot or something?

rockstardev replied:
Sections no longer separated visually ;( 38ca0accde6cf57b40b1bd0c97e8abaab5dfcf38

bolatovumar commented on BTCPayServer.Client/Models/StoreBaseData.cs:
```diff
@@ -30,31 +30,36 @@ public abstract class StoreBaseData
         [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
         public int RecommendedFeeBlockTarget { get; set; } = 1;
 
+        //
```
Same here.

rockstardev replied:
38ca0accde6cf57b40b1bd0c97e8abaab5dfcf38

rockstardev commented:
Two open issues:

 1. "`NBitcoin.Network.Main.GetBech32Encoder(Bech32Type.WITNESS_PUBKEY_ADDRESS, false).HumanReadablePart` instead of my l33t specification of exact strings" -> I'm OK with changing that
 2. "I am not comfortable with it. Or maybe we can just activate it for people checking OnChainWithLnInvoiceFallback so we can get feedback from early users" -> we already had `BITCOIN:` schema previously, but removed it because of QR Code readers would sometimes mess up casing. In the light of all the information @shesek is presenting I think it's best we go back to `BITCOIN:`. Same for uppercase BECH32 address in QR. it's all lowercase otherwise.

If you @NicolasDorier  and @Kukks are giving me yes on both - then I'll make the changes and merge PR so we can get this to @pavlenex for release.

NicolasDorier commented:
@rockstardev ah yeah you are right. Actually now I am surprised we have gone back to lowercase, thinking about it I would have sworn we were lower case for bech32 already...

rockstardev commented:
Merging after addressing all discussion points. Feel free to further tweak in master if you want, but keep in mind that it's rockstar code, so it's already perfect.

pavlenex commented:
Eh uppercase schema `BITCOIN`apparently broke some things #2099
Who should we throw tomatoes at?

rockstardev commented:
@shesek, obviously 😉

shesek commented:
lol, okay, I can accept that... though I think that at least one or two should be thrown at CashApp's direction too :p

Are you certain that the uppercase protocol is indeed the cause? I would wait for #2099's OP (or some other CashApp user) to reply to https://github.com/btcpayserver/btcpayserver/issues/2099#issuecomment-738456041 before concluding this.

If CashApp indeed doesn't support this, I think that its worth trying to reach out them and ask them to fix it. It's causing them to be incompatible with other wallet software too.

Also, the fact that there's only one uppercase-related bug report for CashApp suggests that other wallets do mostly handle this properly.

pavlenex commented:
User was using P2SH btw.
